Additions for evolutions 2

nilsdeppe opened the pull request:
## Proposed changes

Some minor additions that are needed to get evolutions working.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [ ] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

nilsdeppe committed:
Add std::remove_cvref from C++20

nilsdeppe committed:
Add metafunction to get the key type from different maplike classes

kidder commented on src/Utilities/TypeTraits.hpp:
```diff
@@ -345,6 +345,20 @@ constexpr bool is_unsigned_v = std::is_unsigned<T>::value;
+/// \ingroup TypeTraits
+template <class T>
+struct remove_cvref {
+  typedef std::remove_cv_t<std::remove_reference_t<T>> type;
```
why is this not a using?

nilsdeppe replied:
Because this is how it is specified in the standard

kidder commented on src/Parallel/CharmMain.cpp:
```diff
@@ -19,6 +19,23 @@ namespace charmxx {
 template <class...>
 using void_t = void;
 
+template <class T>
```
this looks general; should it be in Utilities instead of Parallel/CharmMain?

nilsdeppe replied:
It could be in theory, I'm not sure if it will be needed elsewhere in practice, though. This is fairly specific to having to register functions with Charm++, in other use cases type deduction will work fine.

wthrowe commented on src/Parallel/CharmMain.cpp:
```diff
@@ -19,6 +19,23 @@ namespace charmxx {
 template <class...>
 using void_t = void;
 
+template <class T>
```
This appears to be only slightly different from just writing `T::value_type` (as implied by the name).  Could you add a comment explaining why it is needed?

nilsdeppe replied:
It's a different type, with a non-const Key, I don't think that needs explanation

wthrowe reviewed (changes requested):
Since we've been picky about it lately: The commit message for your second commit implies it is a pure rename commit, but in fact it also changes other code.  Split it or fix the commit message.

nilsdeppe commented:
No, I'm not changing the commit message. It's a static_assert.

nilsdeppe commented:
I changed my mind on both issues, will address and push updates! (I agree with Will)

nilsdeppe committed:
Use ReceiveDataType instead of ReceiveData_t in Algorithm

nilsdeppe committed:
Add static_assert to Algorithm::receive_data_impl

The assert checks that the data being sent is correct and gives a more
useful error than the normal template backtrace into the STL.

nilsdeppe commented:
I've addressed @wthrowe's comments. Until I get a reply from @kidder I'm not sure what to do on his.
